zypp/repo/yum/Downloader.cc

```diff
@@ -30,109 +28,119 @@ namespace repo
 {
 namespace yum
 {
-
-Downloader::Downloader( const RepoInfo &repoinfo , const Pathname &delta_dir)
-  : repo::Downloader(repoinfo), _delta_dir(delta_dir), _media_ptr(0L)
-{}
-
-RepoStatus Downloader::status( MediaSetAccess &media )
-{
-  RepoStatus ret( media.provideOptionalFile( repoInfo().path() / "/repodata/repomd.xml" ) );
-  if ( !ret.empty() )	// else: mandatory master index is missing
-    ret = ret && RepoStatus( media.provideOptionalFile( "/media.1/media" ) );
-  // else: mandatory master index is missing -> stay empty
-  return ret;
-}
-
-static OnMediaLocation loc_with_path_prefix( const OnMediaLocation & loc, const Pathname & prefix )
-{
-  if (prefix.empty() || prefix == "/")
-    return loc;
-
-  OnMediaLocation loc_with_path(loc);
-  loc_with_path.changeFilename(prefix / loc.filename());
-  return loc_with_path;
-}
-
-// search old repository file file to run the delta algorithm on
-static Pathname search_deltafile( const Pathname & dir, const Pathname & file )
-{
-  Pathname deltafile;
-  if (!PathInfo(dir).isDir())
-    return deltafile;
-  string base = file.basename();
-  size_t hypoff = base.find("-");
-  if (hypoff != string::npos)
-    base.replace(0, hypoff + 1, "");
-  size_t basesize = base.size();
-  std::list<Pathname> retlist;
-  if (!filesystem::readdir(retlist, dir, false))
+  ///////////////////////////////////////////////////////////////////
+  namespace
   {
-    for_( it, retlist.begin(), retlist.end() )
+    inline OnMediaLocation loc_with_path_prefix( OnMediaLocation loc_r, const Pathname & prefix_r )
     {
-      string fn = it->asString();
-      if (fn.size() >= basesize && fn.substr(fn.size() - basesize, basesize) == base)
-	deltafile = *it;
+      if ( ! prefix_r.empty() && prefix_r != "/" )
+	loc_r.changeFilename( prefix_r / loc_r.filename() );
+      return loc_r;
     }
-  }
-  return deltafile;
-}
 
-bool Downloader::patches_Callback( const OnMediaLocation & loc_r, const string & id_r )
-{
-  OnMediaLocation loc_with_path(loc_with_path_prefix(loc_r, repoInfo().path()));
-  MIL << id_r << " : " << loc_with_path << endl;
-  this->enqueueDigested(loc_with_path,  FileChecker(), search_deltafile(_delta_dir + "repodata", loc_r.filename()));
-  return true;
-}
-
-
-//bool repomd_Callback2( const OnMediaLocation &loc, const ResourceType &dtype, const std::string &typestr, UserData & userData_r );
+    // search old repository file file to run the delta algorithm on
+    Pathname search_deltafile( const Pathname & dir, const Pathname & file )
+    {
+      Pathname deltafile;
+      if ( ! PathInfo(dir).isDir() )
+	return deltafile;
+
+      std::string base = file.basename();
+      size_t hypoff = base.find( "-" );
+      if ( hypoff != std::string::npos )
+	base.replace( 0, hypoff + 1, "" );
```

Comment:
why is it required to replace the string before the first hyphen in the filename?

Reply:
`6eff3456ea6b110cb95e825b91d6b62b6e83bd6ab074649baec7f7831fda646e-primary.xml`
Strips the checksum preceding the file stem so we can look for an old `*-primary.xml` which may contain some reusable blocks.